File: 2021/sources/day4.js
const fs = require('fs')
const path = require('path')
let input = fs.readFileSync(path.resolve(__dirname, '../inputs/day4.txt'), 'utf8').split('\n\n')

let drawnNumbers = input.shift().split(',').map(n=>parseInt(n, 10))
let boards = input.map(boardInput => 
	boardInput.split('\n').map(row =>
		row.split(' ')
			.map(number=> parseInt(number.trim(), 10))
			.filter(number => !isNaN(number))
	)
)

const checkRows = (board) => board.reduce((bingo, row) => bingo || 0 == row.reduce((sum, number) => sum + number, 0), false)
const checkColumns = (board) => [0,1,2,3,4].reduce((bingo, column) => bingo || 0 == board.reduce((sum, row) => sum + row[column], 0), false)

let winningBoard = null
let loosingNumber = null
let loosingBoard = null

drawnNumbers.forEach(drawnNumber => {
	boards = boards.map((board, index) => {
		board = board.map(row => row.map(number => number===drawnNumber ? 0 : number))
		if (checkRows(board) || checkColumns(board)) {
			if (!winningBoard) {
				winningBoard = board.reduce((sum, row) => sum + row.reduce((sum, number) => sum + number, 0), 0)
				console.log("BINGO", winningBoard*drawnNumber)
			}
			loosingNumber = drawnNumber
			loosingBoard = board.reduce((sum, row) => sum + row.reduce((sum, number) => sum + number, 0), 0)
			board = [[1000]]
		}
		return board
	})
})

console.log("BONGO", loosingNumber, loosingBoard, loosingBoard*loosingNumber)

File: 2021/sources/day4.swift
import Foundation

public func day4() {
    func toUInt8(_ value: String) -> UInt8 {UInt8(value)!}

    class BingoNumber {
        var value: UInt8 = 0
        var marked: Bool = false
        init(with value: UInt8) {self.value = value}
    }

    struct BingoRow {
        var numbers: [BingoNumber]
        var count: UInt8 { get {numbers.reduce(0) {$0 + $1.value}} }
        mutating func mark(number: UInt8) -> Bool {
            numbers.reduce(false){
                if $1.value == number {$1.marked = true; return true}
                return $0
            }
        }
    }

    struct BingoBoard {
        var rows: [BingoRow] = []
        init(with numbers: [[UInt8]]) {
            rows += numbers.map{BingoRow(numbers: $0.map{BingoNumber(with: $0)})}
            rows += Array(0...4).map{y in BingoRow(numbers: Array(0...4).map{x in BingoNumber(with: numbers[y][x])})}
        }
        mutating func mark(number: UInt8) -> Bool {
            return rows.enumerated().reduce(false) {result, enumeration in
                var row = rows[enumeration.offset]
                let bingo = row.mark(number: number)
                print(row.count)
                return row.count==5 || result
            }
        }
    }

    var input = try! String(contentsOfFile: Bundle.main.path(forResource: "day4", ofType: "txt")!)
        .components(separatedBy: "\n\n")

    let drawnNumbers = input.removeFirst().components(separatedBy: ",").map{UInt8($0)!}
    print(drawnNumbers)

    var boards = input.map{ bingoInput -> BingoBoard in
        let boardInput: [[UInt8]] = bingoInput
            .components(separatedBy: .newlines)
            .map{ lines -> [UInt8] in
                return lines.components(separatedBy: .whitespaces).filter{!$0.isEmpty}.map(toUInt8)
            }
        return BingoBoard(with: boardInput)
    }

    drawnNumbers.forEach { drawnNumber in
        if let winningBoard = boards.enumerated().filter({boards[$0.offset].mark(number: drawnNumber)}).first {
            print("BINGO!")
            let board = winningBoard.element
            print("thats a BINGO!", board.rows[0...4].map{$0.numbers.map{($0.value,$0.marked)}})
            let sum = board.rows[0...4].reduce(UInt8(0)) {partialSum, row in
                print(partialSum)
                return partialSum + row.numbers.reduce(UInt8(0)) {$0 + ($1.marked ? 0 : $1.value)}
            }
            print(sum * drawnNumber)
        }
    }

}
